File: js/src/i18n/icu.ts
/**
 * The subset of ICU MessageFormat the message files use.
 *
 * `Intl.PluralRules` supplies the CLDR categories, so a plural is chosen by the
 * locale's own rules and not by a `=== 1` written here. Node ships full ICU, so
 * every locale it knows has its own categories
 */

type Argument = string | number | boolean | null | undefined;

/** Named arguments a pattern reads */
export type Values = Record<string, Argument>;

type Node =
    | { kind: 'text'; text: string }
    | { kind: 'argument'; name: string; style: NumberStyle | null }
    | { kind: 'hash' }
    | { kind: 'plural'; name: string; branches: Map<string, Node[]>; exact: Map<number, Node[]> }
    | { kind: 'select'; name: string; branches: Map<string, Node[]> };

/** `integer`, or a skeleton naming how many fraction digits to print */
type NumberStyle = { kind: 'integer' } | { kind: 'fraction'; digits: number };

/** A pattern that will not parse, so the caller can print it unformatted */
export class PatternError extends Error {}

const cache = new Map<string, Node[]>();

/**
 * One pattern, with its arguments filled in.
 *
 * @throws PatternError The pattern is not MessageFormat this can read.
 */
export function format(locale: string, pattern: string, values: Values = {}): string {
    let nodes = cache.get(pattern);

    if (nodes === undefined) {
        nodes = parse(pattern);
        cache.set(pattern, nodes);
    }

    return render(nodes, locale, values, null);
}

/** Whether a pattern parses at all, for the test that holds the file to its job */
export function parses(pattern: string): boolean {
    try {
        parse(pattern);

        return true;
    } catch {
        return false;
    }
}

/**
 * Every argument name a pattern reads, so a caller can tell a missing value from
 * a pattern that never wanted one
 */
export function argumentsOf(pattern: string): string[] {
    const names = new Set<string>();
    const walk = (nodes: Node[]): void => {
        for (const node of nodes) {
            if (node.kind === 'argument' || node.kind === 'plural' || node.kind === 'select') {
                names.add(node.name);
            }

            if (node.kind === 'plural') {
                for (const branch of node.exact.values()) {
                    walk(branch);
                }
            }

            if (node.kind === 'plural' || node.kind === 'select') {
                for (const branch of node.branches.values()) {
                    walk(branch);
                }
            }
        }
    };

    walk(parse(pattern));

    return [...names];
}

class Reader {
    public at = 0;

    constructor(public readonly source: string) {}

    peek(offset = 0): string {
        return this.source[this.at + offset] ?? '';
    }

    done(): boolean {
        return this.at >= this.source.length;
    }
}

function parse(pattern: string): Node[] {
    const reader = new Reader(pattern);
    const nodes = readMessage(reader, false);

    if (!reader.done()) {
        throw new PatternError(`Unexpected "${reader.peek()}" at ${reader.at}.`);
    }

    return nodes;
}

/**
 * Text up to the end of the pattern, or to the `}` that closes the branch this
 * is inside. `#` is an argument only inside a plural branch
 */
function readMessage(reader: Reader, nested: boolean): Node[] {
    const nodes: Node[] = [];
    let text = '';

    const flush = (): void => {
        if (text !== '') {
            nodes.push({ kind: 'text', text });
            text = '';
        }
    };

    while (!reader.done()) {
        const char = reader.peek();

        if (char === '}' && nested) {
            break;
        }

        if (char === '{') {
            flush();
            nodes.push(readArgument(reader));

            continue;
        }

        if (char === '#' && nested) {
            flush();
            nodes.push({ kind: 'hash' });
            reader.at++;

            continue;
        }

        if (char === "'") {
            text += readQuoted(reader);

            continue;
        }

        text += char;
        reader.at++;
    }

    flush();

    return nodes;
}

/**
 * The ICU apostrophe rules, which are what a translator gets wrong first.
 *
 * `''` is one apostrophe. A lone `'` opens a quoted run only where it stands in
 * front of a `{`, `}`, `#` or `|`, so `the SDK's` needs no escaping. Inside an
 * open quote `''` is again one apostrophe and does not close it, which is why a
 * run of braces has to be quoted together as `'}}'` and not as `'}''}'`
 */
function readQuoted(reader: Reader): string {
    if (reader.peek(1) === "'") {
        reader.at += 2;

        return "'";
    }

    if (!'{}#|'.includes(reader.peek(1)) || reader.peek(1) === '') {
        reader.at++;

        return "'";
    }

    reader.at++;
    let text = '';

    while (!reader.done()) {
        if (reader.peek() === "'") {
            if (reader.peek(1) === "'") {
                text += "'";
                reader.at += 2;

                continue;
            }

            reader.at++;

            return text;
        }

        text += reader.peek();
        reader.at++;
    }

    // An unterminated quote runs to the end of the pattern, which is what ICU does.
    return text;
}

function readArgument(reader: Reader): Node {
    reader.at++;
    skipSpace(reader);
    const name = readName(reader);
    skipSpace(reader);

    if (reader.peek() === '}') {
        reader.at++;

        return { kind: 'argument', name, style: null };
    }

    expect(reader, ',');
    skipSpace(reader);
    const type = readName(reader);
    skipSpace(reader);

    if (type === 'plural') {
        expect(reader, ',');

        return readPlural(reader, name);
    }

    if (type === 'select') {
        expect(reader, ',');

        return readSelect(reader, name);
    }

    if (type !== 'number') {
        throw new PatternError(`"${type}" is not an argument type this reads.`);
    }

    if (reader.peek() === '}') {
        reader.at++;

        return { kind: 'argument', name, style: null };
    }

    expect(reader, ',');
    skipSpace(reader);
    const style = readStyle(reader);
    skipSpace(reader);
    expect(reader, '}');

    return { kind: 'argument', name, style };
}

/**
 * `integer`, or the fraction-digit part of a number skeleton.
 *
 * `::.00` is two places, `::.` is none. The rest of the skeleton syntax is not
 * read, because a pattern using it would format differently here from the
 * implementation it was written against
 */
function readStyle(reader: Reader): NumberStyle {
    let raw = '';

    while (!reader.done() && reader.peek() !== '}') {
        raw += reader.peek();
        reader.at++;
    }

    raw = raw.trim();

    if (raw === 'integer') {
        return { kind: 'integer' };
    }

    if (raw.startsWith('::')) {
        const skeleton = raw.slice(2).trim();
        const fraction = /^\.(0*)$/.exec(skeleton);

        if (fraction !== null) {
            return { kind: 'fraction', digits: (fraction[1] ?? '').length };
        }
    }

    throw new PatternError(`"${raw}" is not a number style this reads.`);
}

function readPlural(reader: Reader, name: string): Node {
    const branches = new Map<string, Node[]>();
    const exact = new Map<number, Node[]>();

    for (;;) {
        skipSpace(reader);

        if (reader.peek() === '}') {
            reader.at++;

            break;
        }

        if (reader.done()) {
            throw new PatternError(`The plural on "${name}" is not closed.`);
        }

        if (reader.peek() === '=') {
            reader.at++;
            const digits = readName(reader);
            skipSpace(reader);
            exact.set(Number(digits), readBranch(reader));

            continue;
        }

        const keyword = readName(reader);
        skipSpace(reader);
        branches.set(keyword, readBranch(reader));
    }

    if (!branches.has('other')) {
        throw new PatternError(`The plural on "${name}" has no "other" branch.`);
    }

    return { kind: 'plural', name, branches, exact };
}

function readSelect(reader: Reader, name: string): Node {
    const branches = new Map<string, Node[]>();

    for (;;) {
        skipSpace(reader);

        if (reader.peek() === '}') {
            reader.at++;

            break;
        }

        if (reader.done()) {
            throw new PatternError(`The select on "${name}" is not closed.`);
        }

        const keyword = readName(reader);
        skipSpace(reader);
        branches.set(keyword, readBranch(reader));
    }

    if (!branches.has('other')) {
        throw new PatternError(`The select on "${name}" has no "other" branch.`);
    }

    return { kind: 'select', name, branches };
}

function readBranch(reader: Reader): Node[] {
    expect(reader, '{');
    const nodes = readMessage(reader, true);
    expect(reader, '}');

    return nodes;
}

function readName(reader: Reader): string {
    let name = '';

    while (!reader.done() && /[^\s{},]/.test(reader.peek())) {
        name += reader.peek();
        reader.at++;
    }

    if (name === '') {
        throw new PatternError(`Expected a name at ${reader.at}.`);
    }

    return name;
}

function skipSpace(reader: Reader): void {
    while (!reader.done() && /\s/.test(reader.peek())) {
        reader.at++;
    }
}

function expect(reader: Reader, char: string): void {
    if (reader.peek() !== char) {
        throw new PatternError(`Expected "${char}" at ${reader.at}, found "${reader.peek()}".`);
    }

    reader.at++;
}

/** `$hash` is the plural argument the branch belongs to, where it is inside one */
function render(nodes: Node[], locale: string, values: Values, hash: number | null): string {
    let out = '';

    for (const node of nodes) {
        out += renderNode(node, locale, values, hash);
    }

    return out;
}

function renderNode(node: Node, locale: string, values: Values, hash: number | null): string {
    switch (node.kind) {
        case 'text':
            return node.text;

        case 'hash':
            // The plural's own value, through the locale's default number
            // format, so a count in the thousands is grouped.
            return hash === null ? '#' : number(locale, hash, null);

        case 'argument':
            return node.name in values
                ? argument(locale, values[node.name], node.style)
                : `{${node.name}}`;

        case 'plural': {
            if (!(node.name in values)) {
                return `{${node.name}}`;
            }

            const value = Number(values[node.name] ?? 0);
            const branch = node.exact.get(value)
                ?? node.branches.get(category(locale, value))
                ?? node.branches.get('other')
                ?? [];

            return render(branch, locale, values, value);
        }

        case 'select': {
            if (!(node.name in values)) {
                return `{${node.name}}`;
            }

            const value = String(values[node.name] ?? '');
            const branch = node.branches.get(value) ?? node.branches.get('other') ?? [];

            return render(branch, locale, values, hash);
        }
    }
}

function argument(locale: string, value: Argument, style: NumberStyle | null): string {
    if (value === null || value === undefined) {
        return '';
    }

    if (style !== null) {
        return number(locale, Number(value), style);
    }

    // An argument with no type is printed, not formatted: `{tokens}` arrives
    // already grouped by its caller, and putting it through a number format
    // again would group it a second time or round its fraction away.
    return typeof value === 'boolean' ? (value ? '1' : '') : String(value);
}

const formatters = new Map<string, Intl.NumberFormat>();

function number(locale: string, value: number, style: NumberStyle | null): string {
    const key = `${locale}\u0000${style === null ? 'plain' : style.kind}\u0000${style !== null && style.kind === 'fraction' ? style.digits : ''}`;
    let formatter = formatters.get(key);

    if (formatter === undefined) {
        formatter = new Intl.NumberFormat(tag(locale), options(style));
        formatters.set(key, formatter);
    }

    return formatter.format(value);
}

/**
 * `halfEven` is ICU's own default, and the two modes disagree on every exact
 * half: a 0.715 reading prints 0.72 under both, and 0.725 prints 0.72 here and
 * 0.73 under the mode JavaScript defaults to
 */
function options(style: NumberStyle | null): Intl.NumberFormatOptions {
    const rounding = { roundingMode: 'halfEven' } as const;

    if (style === null) {
        return { ...rounding, maximumFractionDigits: 3 };
    }

    if (style.kind === 'integer') {
        return { ...rounding, maximumFractionDigits: 0 };
    }

    return { ...rounding, minimumFractionDigits: style.digits, maximumFractionDigits: style.digits };
}

const rules = new Map<string, Intl.PluralRules>();

function category(locale: string, value: number): string {
    let rule = rules.get(locale);

    if (rule === undefined) {
        rule = new Intl.PluralRules(tag(locale));
        rules.set(locale, rule);
    }

    return rule.select(value);
}

/**
 * A locale as the message files name it, as a BCP 47 tag.
 *
 * The files are named `pt_BR` and `Intl` takes `pt-BR`. A tag `Intl` cannot read
 * falls back to English, because throwing here would take a report down over the
 * language it was going to be printed in
 */
function tag(locale: string): string {
    const bcp = locale.replace('_', '-');

    try {
        return Intl.getCanonicalLocales(bcp)[0] ?? 'en';
    } catch {
        return 'en';
    }
}

File: python/src/jevlint/icu.py
"""
The subset of ICU MessageFormat the message files use.

Babel supplies the CLDR categories and number patterns, so every locale it has
data for brings its own.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from decimal import Decimal
from typing import Any

from babel import Locale, UnknownLocaleError
from babel.numbers import format_decimal

Values = dict[str, Any]


class PatternError(Exception):
    """A pattern that will not parse, so the caller can print it unformatted."""


@dataclass(frozen=True)
class NumberStyle:
    """`integer`, or a skeleton naming how many fraction digits to print."""

    kind: str

    digits: int = 0


@dataclass
class Node:
    kind: str

    text: str = ''

    name: str = ''

    style: NumberStyle | None = None

    branches: dict[str, list[Node]] = field(default_factory=dict)

    exact: dict[float, list[Node]] = field(default_factory=dict)


_parsed: dict[str, list[Node]] = {}


def format(locale: str, pattern: str, values: Values | None = None) -> str:
    """
    One pattern, with its arguments filled in.

    Raises PatternError where the pattern is not MessageFormat this can read.
    """
    nodes = _parsed.get(pattern)

    if nodes is None:
        nodes = parse(pattern)
        _parsed[pattern] = nodes

    return _render(nodes, locale, values or {}, None)


def parses(pattern: str) -> bool:
    """Whether a pattern parses at all, for the test that holds the file to its job."""
    try:
        parse(pattern)
    except PatternError:
        return False

    return True


def arguments_of(pattern: str) -> list[str]:
    """
    Every argument name a pattern reads, so a caller can tell a missing value
    from a pattern that never wanted one.
    """
    names: list[str] = []

    def walk(nodes: list[Node]) -> None:
        for node in nodes:
            if node.kind in ('argument', 'plural', 'select') and node.name not in names:
                names.append(node.name)

            for branch in list(node.exact.values()) + list(node.branches.values()):
                walk(branch)

    walk(parse(pattern))

    return names


class _Reader:
    def __init__(self, source: str) -> None:
        self.source = source
        self.at = 0

    def peek(self, offset: int = 0) -> str:
        position = self.at + offset

        return self.source[position] if 0 <= position < len(self.source) else ''

    def done(self) -> bool:
        return self.at >= len(self.source)


def parse(pattern: str) -> list[Node]:
    reader = _Reader(pattern)
    nodes = _read_message(reader, False)

    if not reader.done():
        raise PatternError(f'Unexpected "{reader.peek()}" at {reader.at}.')

    return nodes


def _read_message(reader: _Reader, nested: bool) -> list[Node]:
    """
    Text up to the end of the pattern, or to the `}` that closes the branch this
    is inside. `#` is an argument only inside a plural branch.
    """
    nodes: list[Node] = []
    text = ''

    def flush() -> None:
        nonlocal text

        if text != '':
            nodes.append(Node('text', text=text))
            text = ''

    while not reader.done():
        char = reader.peek()

        if char == '}' and nested:
            break

        if char == '{':
            flush()
            nodes.append(_read_argument(reader))

            continue

        if char == '#' and nested:
            flush()
            nodes.append(Node('hash'))
            reader.at += 1

            continue

        if char == "'":
            text += _read_quoted(reader)

            continue

        text += char
        reader.at += 1

    flush()

    return nodes


def _read_quoted(reader: _Reader) -> str:
    """
    The ICU apostrophe rules, which are what a translator gets wrong first.

    `''` is one apostrophe. A lone `'` opens a quoted run only where it stands in
    front of a `{`, `}`, `#` or `|`, so `the SDK's` needs no escaping. Inside an
    open quote `''` is again one apostrophe and does not close it, which is why a
    run of braces has to be quoted together as `'}}'` and not as `'}''}'`.
    """
    if reader.peek(1) == "'":
        reader.at += 2

        return "'"

    if reader.peek(1) not in ('{', '}', '#', '|'):
        reader.at += 1

        return "'"

    reader.at += 1
    text = ''

    while not reader.done():
        if reader.peek() == "'":
            if reader.peek(1) == "'":
                text += "'"
                reader.at += 2

                continue

            reader.at += 1

            return text

        text += reader.peek()
        reader.at += 1

    # An unterminated quote runs to the end of the pattern, which is what ICU does.
    return text


def _read_argument(reader: _Reader) -> Node:
    reader.at += 1
    _skip_space(reader)
    name = _read_name(reader)
    _skip_space(reader)

    if reader.peek() == '}':
        reader.at += 1

        return Node('argument', name=name)

    _expect(reader, ',')
    _skip_space(reader)
    kind = _read_name(reader)
    _skip_space(reader)

    if kind == 'plural':
        _expect(reader, ',')

        return _read_plural(reader, name)

    if kind == 'select':
        _expect(reader, ',')

        return _read_select(reader, name)

    if kind != 'number':
        raise PatternError(f'"{kind}" is not an argument type this reads.')

    if reader.peek() == '}':
        reader.at += 1

        return Node('argument', name=name)

    _expect(reader, ',')
    _skip_space(reader)
    style = _read_style(reader)
    _skip_space(reader)
    _expect(reader, '}')

    return Node('argument', name=name, style=style)


def _read_style(reader: _Reader) -> NumberStyle:
    """
    `integer`, or the fraction-digit part of a number skeleton.

    `::.00` is two places, `::.` is none. The rest of the skeleton syntax is not
    read, because a pattern using it would format differently here from the
    implementation it was written against.
    """
    raw = ''

    while not reader.done() and reader.peek() != '}':
        raw += reader.peek()
        reader.at += 1

    raw = raw.strip()

    if raw == 'integer':
        return NumberStyle('integer')

    if raw.startswith('::'):
        skeleton = raw[2:].strip()

        if skeleton.startswith('.') and set(skeleton[1:]) <= {'0'}:
            return NumberStyle('fraction', len(skeleton) - 1)

    raise PatternError(f'"{raw}" is not a number style this reads.')


def _read_plural(reader: _Reader, name: str) -> Node:
    node = Node('plural', name=name)

    while True:
        _skip_space(reader)

        if reader.peek() == '}':
            reader.at += 1

            break

        if reader.done():
            raise PatternError(f'The plural on "{name}" is not closed.')

        if reader.peek() == '=':
            reader.at += 1
            digits = _read_name(reader)
            _skip_space(reader)
            node.exact[float(digits)] = _read_branch(reader)

            continue

        keyword = _read_name(reader)
        _skip_space(reader)
        node.branches[keyword] = _read_branch(reader)

    if 'other' not in node.branches:
        raise PatternError(f'The plural on "{name}" has no "other" branch.')

    return node


def _read_select(reader: _Reader, name: str) -> Node:
    node = Node('select', name=name)

    while True:
        _skip_space(reader)

        if reader.peek() == '}':
            reader.at += 1

            break

        if reader.done():
            raise PatternError(f'The select on "{name}" is not closed.')

        keyword = _read_name(reader)
        _skip_space(reader)
        node.branches[keyword] = _read_branch(reader)

    if 'other' not in node.branches:
        raise PatternError(f'The select on "{name}" has no "other" branch.')

    return node


def _read_branch(reader: _Reader) -> list[Node]:
    _expect(reader, '{')
    nodes = _read_message(reader, True)
    _expect(reader, '}')

    return nodes


def _read_name(reader: _Reader) -> str:
    name = ''

    while not reader.done() and not reader.peek().isspace() and reader.peek() not in ('{', '}', ','):
        name += reader.peek()
        reader.at += 1

    if name == '':
        raise PatternError(f'Expected a name at {reader.at}.')

    return name


def _skip_space(reader: _Reader) -> None:
    while not reader.done() and reader.peek().isspace():
        reader.at += 1


def _expect(reader: _Reader, char: str) -> None:
    if reader.peek() != char:
        raise PatternError(f'Expected "{char}" at {reader.at}, found "{reader.peek()}".')

    reader.at += 1


def _render(nodes: list[Node], locale: str, values: Values, hash_value: float | None) -> str:
    """`hash_value` is the plural argument the branch belongs to, where it is inside one."""
    return ''.join(_render_node(node, locale, values, hash_value) for node in nodes)


def _render_node(node: Node, locale: str, values: Values, hash_value: float | None) -> str:
    if node.kind == 'text':
        return node.text

    if node.kind == 'hash':
        # The plural's own value, through the locale's default number format, so
        # a count in the thousands is grouped.
        return '#' if hash_value is None else _number(locale, hash_value, None)

    if node.kind == 'argument':
        if node.name not in values:
            return '{' + node.name + '}'

        return _argument(locale, values[node.name], node.style)

    if node.kind == 'plural':
        if node.name not in values:
            return '{' + node.name + '}'

        value = _as_number(values[node.name])
        branch = node.exact.get(value)

        if branch is None:
            branch = node.branches.get(_category(locale, value), node.branches.get('other', []))

        return _render(branch, locale, values, value)

    if node.name not in values:
        return '{' + node.name + '}'

    chosen = '' if values[node.name] is None else str(values[node.name])
    branch = node.branches.get(chosen, node.branches.get('other', []))

    return _render(branch, locale, values, hash_value)


def _argument(locale: str, value: Any, style: NumberStyle | None) -> str:
    if value is None:
        return ''

    if style is not None:
        return _number(locale, _as_number(value), style)

    # An argument with no type is printed, not formatted: `{tokens}` arrives
    # already grouped by its caller, and putting it through a number format again
    # would group it a second time or round its fraction away.
    if isinstance(value, bool):
        return '1' if value else ''

    if isinstance(value, float) and value.is_integer():
        return str(int(value))

    return str(value)


def _as_number(value: Any) -> float:
    if isinstance(value, bool):
        return 1.0 if value else 0.0

    if isinstance(value, (int, float)):
        return float(value)

    try:
        return float(value)
    except (TypeError, ValueError):
        return 0.0


_locales: dict[str, Locale] = {}


def _locale(locale: str) -> Locale:
    """
    A locale as the message files name it, as CLDR names it.

    A tag Babel has no data for falls back to English, because raising here would
    take a report down over the language it was going to be printed in.
    """
    found = _locales.get(locale)

    if found is None:
        try:
            found = Locale.parse(locale.replace('-', '_'))
        except (UnknownLocaleError, ValueError, TypeError):
            found = Locale.parse('en')

        _locales[locale] = found

    return found


def _number(locale: str, value: float, style: NumberStyle | None) -> str:
    """
    CLDR's own number format, which rounds half to even.

    The two modes disagree on every exact half: a 0.715 reading prints 0.72 under
    both, and 0.725 prints 0.72 here and 0.73 under half away from zero.
    """
    if style is None:
        pattern = '#,##0.###'
    elif style.kind == 'integer':
        pattern = '#,##0'
    else:
        pattern = '#,##0.' + '0' * style.digits if style.digits > 0 else '#,##0'

    return format_decimal(Decimal(repr(value)), format=pattern, locale=_locale(locale))


def _category(locale: str, value: float) -> str:
    return _locale(locale).plural_form(int(value) if float(value).is_integer() else value)
